### PR TITLE
Add service plan and parameter completion

### PR DESCRIPTION
**Pull request by geoand**
## What is the purpose of this change? What does it change?
See $SUBJECT

## Was the change discussed in an issue?
Fixes: #1171

## How to test changes?
Assuming that the completion scripts have already been setup and that the binary from this PR is on $PATH, a good test would be the following:

Now by execuring

`odo service create $some_service --plan + [TAB]` 

you should yield the various plans

Once you decide on a plan, you can see the required parameters by using

`odo service create $some_service --plan $some_plan -p + [TAB]`

### CODE REVIEW

File: scripts/generate-coverage.sh [hunk 1]
```diff
@@ -6,7 +6,7 @@
 set -e
 echo "" > coverage.txt
 go test -i -race ./cmd/odo
-for d in $(go list ./... | grep -v vendor | grep -v tests/e2e); do
+for d in $(go list ./... | grep -v vendor | grep -v tests/e2e | grep -v testingutil); do
```
Comment (metacosm) [inline scripts/generate-coverage.sh]: Sneaky! :)

File: pkg/odo/cli/service/service.go [hunk 2]
```diff
@@ -47,8 +46,5 @@ func NewCmdService(name, fullName string) *cobra.Command {
 	appCmd.AddApplicationFlag(serviceDeleteCmd)
 	appCmd.AddApplicationFlag(serviceListCmd)
 
-	completion.RegisterCommandHandler(serviceCreateCmd, completion.ServiceClassCompletionHandler)
-	completion.RegisterCommandHandler(serviceDeleteCmd, completion.ServiceCompletionHandler)
```
Comment (metacosm) [inline pkg/odo/cli/service/service.go]: Why did you remove these?

File: scripts/generate-coverage.sh (hunk 1 above)
Comment (geoand) [reply scripts/generate-coverage.sh]: Unfortunately it didn't have the effect I had hoped for :)

File: pkg/odo/cli/service/service.go (hunk 2 above)
Comment (geoand) [reply pkg/odo/cli/service/service.go]: Because they already exist in the respective command handlers, so these were redundant

Comment (metacosm) [reply pkg/odo/cli/service/service.go]: Was wondering about this. Cool!

File: pkg/odo/util/completion/completionhandlers_test.go [hunk 3]
```diff
@@ -18,6 +22,239 @@ import (
...
+	classExternalMetaData := make(map[string]interface{})
+	classExternalMetaDataRaw, err := json.Marshal(classExternalMetaData)
+	if err != nil {
+		fmt.Printf("error occured %v during marshalling", err)
+		return
+	}
+
+	serviceClassList := &scv1beta1.ClusterServiceClassList{
+		Items: []scv1beta1.ClusterServiceClass{
+			{
```
[the remark this answers is not in the document]
Comment (geoand) [reply pkg/odo/util/completion/completionhandlers_test.go]: I must have missed that bad body :)

[the remark this answers is not in the document]
Comment (geoand) [reply pkg/odo/util/completion/completionhandlers_test.go]: Fixed

File: pkg/odo/util/completion/completion.go [hunk 4]
```diff
@@ -48,7 +48,24 @@ func NewParsedArgs(args complete.Args, cmd *cobra.Command) parsedArgs {
...
 	commands, flagValues := getCommandsAndFlags(typed, cmd)
 
-	complete.Log("Parsed flag values: %v", flagValues)
+	complete.Log("Parsed commands values for full input: %v", commands)
+	complete.Log("Parsed flag values for full input: %v", flagValues)
+
+	// this part is meant to handle incomplete flags when other input already exists,
+	// so 2 tokens is the minimum that makes sense
+	if len(typed) > 2 {
+		typedWithoutLast := make([]string, len(typed))
```
Comment (metacosm) [inline pkg/odo/util/completion/completion.go]: Could you comment the logic here some more, please?

Comment (geoand) [reply pkg/odo/util/completion/completion.go]: Sure thing

Comment (geoand) [reply pkg/odo/util/completion/completion.go]: Done

**Comment by mik-dass** [comment]
It seems that ServiceCompletionHandler and ServiceClassCompletionHandler were never updated to fix the duplication (press <tab> twice, the suggestions appear twice). The other commands have been updated. Should we fix that in this PR itself?

**Comment by geoand** [comment]
> It seems that ServiceCompletionHandler and ServiceClassCompletionHandler were never updated to fix the duplication (press twice, the suggestions appear twice). The other commands have been updated. Should we fix that in this PR itself?

I would say let's do that in another PR.
I am volunteering you @mik-dass :P :P :P

**Comment by geoand** [comment]
@metacosm Caching the completion stuff is definitely something to look into the future

**Comment by mik-dass** [comment]
> Makes sense. Probably best to open an issue then and have it depend on this :). Once this is merged, then it would make sense for you to pick this up

But I have already raised a PR #1214 :D Was a very basic fix

File: pkg/odo/util/completion/completionhandlers.go [hunk 5]
```diff
@@ -39,16 +40,92 @@ var ServiceClassCompletionHandler = func(cmd *cobra.Command, args parsedArgs, co
...
+	if len(args.original.Completed) < 2 {
+		complete.Log("Couldn't extract the service name")
+		return completions
+	}
+
+	inputServiceName := args.original.Completed[1]
+
+	complete.Log(fmt.Sprintf("Using input: serviceName = %s", inputServiceName))
+
+	_, servicePlans, err := service.GetServiceClassAndPlans(context.Client, inputServiceName)
```
Comment (metacosm) [inline pkg/odo/util/completion/completionhandlers.go]: Actually, maybe use https://github.com/redhat-developer/odo/blob/master/pkg/occlient/occlient.go#L1955 instead?

File: pkg/odo/util/completion/completionhandlers.go [hunk 6]
```diff
@@ -39,16 +40,92 @@ var ServiceClassCompletionHandler = func(cmd *cobra.Command, args parsedArgs, co
...
+	inputServiceName := args.original.Completed[1]
+	inputPlanName := args.flagValues["plan"]
+
+	complete.Log(fmt.Sprintf("Using input: serviceName = %s, servicePlan = %s", inputServiceName, inputPlanName))
+
+	_, servicePlans, err := service.GetServiceClassAndPlans(context.Client, inputServiceName)
+	if err != nil {
+		complete.Log("Error retrieving details of service")
+		return completions
+	}
```
Comment (metacosm) [inline pkg/odo/util/completion/completionhandlers.go]: Can we try to share some code between the 2 completion handlers?

File: pkg/odo/util/completion/completionhandlers.go (hunk 5 above)
Comment (geoand) [reply pkg/odo/util/completion/completionhandlers.go]: I'll look into it

File: pkg/odo/util/completion/completionhandlers.go (hunk 6 above)
Comment (geoand) [reply pkg/odo/util/completion/completionhandlers.go]: Checking

Comment (geoand) [reply pkg/odo/util/completion/completionhandlers.go]: @metacosm I don't think this makes sense. 
Checking the updated handlers there is not much in common

File: pkg/odo/util/completion/completionhandlers.go (hunk 5 above)
Comment (geoand) [reply pkg/odo/util/completion/completionhandlers.go]: Fixed

**Comment by geoand** [comment]
@mik-dass Have you reviewed the PR? If you're OK with it feel free to merge

File: pkg/odo/util/completion/completionhandlers.go [hunk 7]
```diff
@@ -39,16 +40,92 @@ var ServiceClassCompletionHandler = func(cmd *cobra.Command, args parsedArgs, co
...
 		completions = append(completions, class.Spec.ExternalName)
 	}
 
 	return
 }
 
+// ServicePlanCompletionHandler provides completion for the the plan of a selected service
+var ServicePlanCompletionHandler = func(cmd *cobra.Command, args parsedArgs, context *genericclioptions.Context) (completions []string) {
+	completions = make([]string, 0)
+	if len(args.original.Completed) < 2 {
```
Comment (mik-dass) [inline pkg/odo/util/completion/completionhandlers.go]: maybe a comment here?

Comment (geoand) [reply pkg/odo/util/completion/completionhandlers.go]: Sure thing

File: pkg/odo/util/completion/completionhandlers.go (hunk 5 above)
Comment (geoand) [reply pkg/odo/util/completion/completionhandlers.go]: So I will add the necessary call for the class

File: pkg/odo/util/completion/completionhandlers.go (hunk 7 above)
Comment (mik-dass) [reply pkg/odo/util/completion/completionhandlers.go]: `the user`?

Comment (geoand) [reply pkg/odo/util/completion/completionhandlers.go]: Fixed

File: pkg/odo/util/completion/completionhandlers_test.go [hunk 8]
```diff
@@ -18,6 +20,201 @@ import (
...
+
+		fakeClientSet.ServiceCatalogClientSet.PrependReactor("list", "clusterserviceplans", func(action ktesting.Action) (handled bool, ret runtime.Object, err error) {
+			return true, &scv1beta1.ClusterServicePlanList{Items: tt.returnedServicePlan}, nil
+		})
+
+		completions := ServicePlanCompletionHandler(nil, tt.parsedArgs, context)
+		sort.Strings(completions)
+
+		// Sort the output and expected o/p in-order to avoid issues due to order as its not important
+		sort.Strings(completions)
```
Comment (mik-dass) [inline pkg/odo/util/completion/completionhandlers_test.go]: why are we sorting twice? I didn't understand from the comment above

Comment (geoand) [reply pkg/odo/util/completion/completionhandlers_test.go]: Damn it... I wanted to sort the expected here. Good catch!

File: pkg/odo/util/completion/completionhandlers_test.go [hunk 9]
```diff
@@ -18,6 +20,201 @@ import (
...
+
+		fakeClientSet.ServiceCatalogClientSet.PrependReactor("list", "clusterserviceplans", func(action ktesting.Action) (handled bool, ret runtime.Object, err error) {
+			return true, &scv1beta1.ClusterServicePlanList{Items: tt.returnedServicePlan}, nil
+		})
+
+		completions := ServiceParameterCompletionHandler(nil, tt.parsedArgs, context)
+		sort.Strings(completions)
+
+		// Sort the output and expected o/p in-order to avoid issues due to order as its not important
+		sort.Strings(completions)
```
Comment (mik-dass) [inline pkg/odo/util/completion/completionhandlers_test.go]: same here as above, why are we sorting twice?

File: pkg/odo/util/completion/completionhandlers_test.go (hunk 8 above)
Comment (geoand) [reply pkg/odo/util/completion/completionhandlers_test.go]: Fixed

File: pkg/odo/util/completion/completionhandlers_test.go (hunk 9 above)
Comment (geoand) [reply pkg/odo/util/completion/completionhandlers_test.go]: Fixed

File: pkg/odo/util/completion/completionhandlers_test.go (hunk 8 above)
Comment (mik-dass) [reply pkg/odo/util/completion/completionhandlers_test.go]: :+1:

File: pkg/odo/util/completion/completionhandlers_test.go (hunk 9 above)
Comment (mik-dass) [reply pkg/odo/util/completion/completionhandlers_test.go]: :+1:
